hermes/bin/src/packaging/wasm_module.rs

```diff
@@ -0,0 +1,115 @@
+//! Wasm module package.
+
+use std::path::Path;
+
+use super::{
+    copy_dir_recursively_to_package, copy_file_from_dir_to_package, DirNotFoundError,
+    FileNotFoundError,
+};
+use crate::errors::Errors;
+
+/// Wasm module package.
+#[derive(Debug)]
+pub(crate) struct WasmModulePackage {
+    /// hdf5 package instance
+    _package: hdf5::File,
+}
+
+impl WasmModulePackage {
+    /// Config JSON file name.
+    const CONFIG_JSON: &'static str = "config.json";
+    /// Config schema JSON file name.
+    const CONFIG_SCHEMA_JSON: &'static str = "config.schema.json";
+    /// Metadata JSON file name.
+    const METADATA_JSON: &'static str = "metadata.json";
+    /// Module wasm file name.
+    const MODULE_WASM: &'static str = "module.wasm";
+    /// Settings schema JSON file name.
+    const SETTINGS_SCHEMA_JSON: &'static str = "settings.schema.json";
+    /// Share directory name.
+    const SHARE: &'static str = "share";
+
+    /// Create a new Wasm module package from a directory.
+    pub(crate) fn from_dir<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
+        let path = path.as_ref();
+        let mut errors = Errors::new();
+
+        let package_path = path.join("module.hdf5");
```

Comment:
As per the design documentation, the file name is to come from the manifest.
And is to have a `.hmod` extension.

I have reconsidered the idea of just copying everything from a directory and I don't like it.
Refer to the design documentation here:
https://input-output-hk.github.io/hermes/architecture/08_concepts/hermes_packaging_requirements/wasm_modules/#overview-of-a-wasm-component-module
The original design is that the manifest file controls how the package is built and is the source of the actual `metadata.json` file.  It is a superset of `metadata.json`.
The packaging process reads the manifest, collects the necessary file and produces the correctly named module.

I think the CLI should just be:
```
hermes package module <manifest_file> [<optional output path>]
```
`package` should have a sub command `module` because we will have multiple `packaging` type commands.
The manifest file is mandatory, and if the output path is not set, the output goes to the current working directory.
We shouldn't need to say `--manifest <file>`  it seems pointless if we always need the parameter, we should use positional arguments.  Same with output path.  `--something` type parameters should be reserved for things which are truly optional.

For now, rename this `.hdf5` extension to `.hmod`
The next PR will correct this, and use the manifest file as originally intended.
It will also then construct the `metadata.json` file not take one from the user, as all the data needed for `metadata.json` is either in the manifest file itself, or is generated by the build process.

Reply:
Updated